src/Collapse/include/gudhi/Flag_complex_edge_collapser.h

```diff
@@ -0,0 +1,375 @@
+    sparse_row_adjacency_matrix_[rw_v].insert(rw_u) = idx;
+    return std::minmax(rw_u, rw_v);
+  }
+
+ public:
+  /** \brief Flag_complex_edge_collapser constructor from a range of filtered edges.
+   *
+   * @param[in] edges Range of Filtered edges range.There is no need the range to be sorted, as it will be performed in
+   * `Flag_complex_edge_collapser::process_edges`.
+   *
+   * \tparam FilteredEdgeRange must be a range for which std::begin and std::end return iterators on a
+   * `Flag_complex_edge_collapser::Filtered_edge`.
+   */
+  template<typename FilteredEdgeRange>
+  Flag_complex_edge_collapser(const FilteredEdgeRange& edges)
+  : f_edge_vector_(std::begin(edges), std::end(edges)) { }
+
+  /** \brief Performs edge collapse in a increasing sequence of the filtration value.
+   *
+   * \tparam filtered_edge_output is a functor that is called on the output edges, in non-decreasing order of
+   * filtration, as filtered_edge_output(u, v, f) where u and v are Vertex_handle representing the extremities of the
+   * edge, and f is its new Filtration_value.
+   */
+  template<typename FilteredEdgeOutput>
+  void process_edges(FilteredEdgeOutput filtered_edge_output) {
+    // Sort edges
+    auto sort_by_filtration = [](const Filtered_edge& edge_a, const Filtered_edge& edge_b) -> bool
+    {
+      return (std::get<2>(edge_a) < std::get<2>(edge_b)); 
+    };
+
+#ifdef GUDHI_USE_TBB
+    tbb::parallel_sort(f_edge_vector_.begin(), f_edge_vector_.end(), sort_by_filtration);
+#else
+    std::sort(f_edge_vector_.begin(), f_edge_vector_.end(), sort_by_filtration);
+#endif
+
+    for (Edge_index endIdx = 0; endIdx < f_edge_vector_.size(); endIdx++) {
+      Filtered_edge fec = f_edge_vector_[endIdx];
+      Vertex_handle u = std::get<0>(fec);
+      Vertex_handle v = std::get<1>(fec);
+      Filtration_value filt = std::get<2>(fec);
+
+      // Inserts the edge in the sparse matrix to update the graph (G_i)
+      IEdge ie = insert_new_edge(u, v, endIdx);
+
+      iedge_to_index_map_.emplace(ie, endIdx);
+      critical_edge_indicator_.push_back(false);
+
+      if (!edge_is_dominated(u, v)) {
+        critical_edge_indicator_[endIdx] = true;
+        filtered_edge_output(u, v, filt);
+        if (endIdx > 1)
+          set_edge_critical(endIdx, filt, filtered_edge_output);
+      }
+    }
+  }
+
+};
+
+/** \brief Constructs a Flag complex from edges as an input, collapse edges and returns remaining edges.
+ *
+ * \fn auto Gudhi::collapse::flag_complex_collapse_edges(FilteredEdgeRange const& edges)
+ *
+ * \tparam FilteredEdgeRange furnishes `std::begin` and `std::end` methods and returns an iterator on a
+ * FilteredEdge of type `std::tuple<Vertex_handle, Vertex_handle, Filtration_value>`
+ *
+ * \ingroup edge_collapse
+ * 
+ */
+template<class FilteredEdgeRange> auto flag_complex_collapse_edges(const FilteredEdgeRange& edges) {
+  auto first_edge_itr = std::begin(edges);
+  auto first_vertex = std::get<0>(*first_edge_itr);
```

Comment:
Trouble if the range is empty?
`using Vertex_handle = std::decay_t<decltype(std::get<0>(*first_edge_itr))>;`
has the advantage that it isn't actually evaluated, only the types are considered.

Reply:
Thank you ! I could not find the right syntax with `std::decay`.
Done on 230bbe96

Reply:
Just in case, an alternative would be `using E = typename std::iterator_traits<decltype(first_edge_itr)>::value_type; using Vertex_handle = std::tuple_element_t<0, E>;`.